server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -262,6 +279,48 @@ protected void doClose() {
         }
     }
 
+    // Inspects all cluster state elements that contain a hint about what the current repository generation is and updates
+    // #latestKnownRepoGen if a newer than currently known generation is found
+    @Override
+    public void applyClusterState(ClusterChangedEvent event) {
+        final ClusterState state = event.state();
+        final String repoName = metadata.name();
+        long bestGenerationFromCS = RepositoryData.EMPTY_REPO_GEN;
+
+        final SnapshotsInProgress snapshotsInProgress = state.custom(SnapshotsInProgress.TYPE);
+        if (snapshotsInProgress != null) {
+            final SnapshotsInProgress.Entry entry = snapshotsInProgress.entries().stream()
+                .filter(e -> e.snapshot().getRepository().equals(repoName)).findFirst()
+                .orElse(null);
+            if (entry != null) {
+                bestGenerationFromCS = entry.getRepositoryStateId();
+            }
+        }
+
+        final SnapshotDeletionsInProgress deletionsInProgress = state.custom(SnapshotDeletionsInProgress.TYPE);
+        if (bestGenerationFromCS == RepositoryData.EMPTY_REPO_GEN && deletionsInProgress != null) {
+            final SnapshotDeletionsInProgress.Entry entry = deletionsInProgress.getEntries().stream()
+                .filter(e -> e.getSnapshot().getRepository().equals(repoName)).findFirst()
+                .orElse(null);
+            if (entry != null) {
+                bestGenerationFromCS = entry.getRepositoryStateId();
+            }
+        }
+
+        final RepositoryCleanupInProgress cleanupInProgress = state.custom(RepositoryCleanupInProgress.TYPE);
```

Comment:
would it make sense for these 3 customs to have entries implement an interface that exposes two methods (repository() and getRepositoryStateId()), which will allow simplifying the code here.

Reply:
Done in https://github.com/elastic/elasticsearch/commit/cfb104f9926f477a6a0fa8a47a7bbd83268f708b ... thanks for this suggestion, this actually will allow simplifying a number of spots in a follow-up where we check all three of these `Custom` values for the presence of a certain repo :)